zegami_sdk/collection.py

```diff
@@ -291,6 +305,104 @@ def imageset_index_to_row_index(self, imageset_index, source=0) -> int:
             raise IndexError(
                 "Invalid imageset index {} for this source"
                 .format(imageset_index))
+    
+    def add_snapshot(self, name, desc, snapshot):
+        url = '{}/{}/project/{}/snapshots/{}/snapshots'.format(
+            self.client.HOME, self.client.API_0, self.workspace_id, self.id)
+        payload = {
+            'name': name,
+            'description': desc,
+            'snapshot': snapshot,
+            'version': 3,
+        }
+        r = self.client._auth_post(url, json.dumps(payload), return_response=True)
+        return r
+
+    def add_feature_pipeline(self, name, steps, source=0, generate_snapshot=False):
+        # get the source
+        source = self._parse_source(source)
+        node_group = [
+            'source_{}'.format(source.name),
+            'collection_{}'.format(self.id),
+        ]
+
+        mRMR_params = steps[0]['params']
+
+        # find the feature extraction node
+        source_feature_extraction_node = self.get_feature_extraction_imageset_id(source)
+
+        join_dataset_id = source._imageset_dataset_join_id
+        imageset_parents = [source_feature_extraction_node]
+        dataset_parents = [self._dataset_id, join_dataset_id]
+
+        mRMR_node = add_node(
+            self.client,
+            self.workspace,
+            action=steps[0]['action'],
```

Comment:
This assumes that the steps will always be ['mrmr', 'cluster'].
I had actually envisaged the user being able to put in an arbitrary sequence of nodes, which is why the original interface was an array.
It's possible that that idea doesn't really work - I guess mRMR output can't be used unless it's fed into a cluster node.
If we are going to make this interface about adding mRMR feeding into UMAP/PCA, then the interface should be explicit about that, e.g.
```
collection.add_feature_pipeline(
  name,
  source,
  mrmr_params=mRMR_params,  # Optional, user could leave this as none and just add a new UMAP
  cluster_params= {
    'algorithm': 'UMAP',
    'n_neighbours': 3,
    etc...
  }
)
```

Another useful use-case to support here later would be PCA -> TSNE. TSNE gives nice results but it is too heavy when there are lots of features, so a PCA step first can help. We haven't added TSNE yet though.

Reply:
I think the tricky thing is if we have any number and type of steps in the pipeline, it's hard to know what the parents are for each node, unless we specify the case of parents for each type of node